docs/proposals/20221611-workload-identity-integration.md

```diff
@@ -0,0 +1,351 @@
+
+The first step for creating K8s cluster using capz is creating a bootstrap cluster which is a Kind cluster and then deploying capi, capz and other helper components. On a high level the workflow looks the following to be able to use AWI and create a management cluster.
+
+**Bootstrap Cluster**
+
+- The operator/admin should be able generate signing key pair or BYO key pair. 
+
+- The bootstrap cluster i.e the Kind K8s cluster should be configured with appropriate flags on kube-apiserver and kube-controller-manager and the key pairs should be mounted on the container path for control plane node. 
+
+- The operator/admin should set up OIDC issuer URL which means having access to azure blob storage where the following two documents will be stored in the blob storage container.
+  - Generate and upload the Discovery Document.
+  - Generate and upload the JWKS Document.
+ 
+- The AWI mutating admission webhook should be installed on the bootstrap cluster.
+
+- A service account should be created in the same namespace as that of capz pod and tagged with the client ID of the Azure AD application or user addigned identity on the bootstrap cluster. The service account should also be associated as a Subject with the appropriate ClusterRoleBindings and RoleBindings for the usual RBAC. This is the service account that the capz pod should use.
+
+- The capz manager service account should be annotated with the Azure AD or User Assigned Client ID. 
+
+- A fedreated credential should be created between the identity and the OIDC issuer URL and the service account(subject). 
+
+- CAPZ pod should use the `azidentity` module from the azure-sdk-for-go to exchange AD token in lieu of the projected service account token.
+
+**Management Cluster**
+
+- Once a K8s cluster is created from the bootstrap cluster, to convert it into a management cluster `clusterctl init` and `clusterctl move` commands are executed.
+
+- After `clusterctl init` command is executed, the control plane node should have the key pairs and then patched to include them in the container path and set the kube-apiserver and kube-controller-manager flags.
+
+- The capz pod service account should be annotated with the client ID of the User Assigned Identity. The client ID can be passed as an argument to the `clusterctl init` command. 
+
+### <a name='ImplementationDetailsNotesConstraints'></a>Implementation Details/Notes/Constraints
+
+- AAD pod identity can co exist with AWI
+- Migration plan from AAD pod identity to AWI for existing cluster. Refer to the `Migration Plan` section at the bottom of the document.
+- For AWI to work the following prerequisites must be met.
+  - Key Generation
+  - OIDC URL Setup
+  - Set Service Account Signing Flags
```

Comment:
These are only required if the management cluster is a kind cluster. For managed clusters, all these are set up as part of cluster bootstrap: https://azure.github.io/azure-workload-identity/docs/installation/managed-clusters.html

Reply:
For managed cluster -- you mean AKS, right? 

But I will add a note about this.

Reply:
> For managed cluster -- you mean AKS, right?

that's right! This [doc](https://azure.github.io/azure-workload-identity/docs/installation/managed-clusters.html) has the list of managed clusters and how to get the oidc URL for setting up federated identity credential.

Reply:
> How does this work for CI when using kind as the management cluster? Can this setup be done once in the CI sub or will we need to do the steps every time? If every time, that requires CI to have creds to run these steps so we will need to authenticate via SP which defeats the purpose of no longer needing SPs to test CAPZ with kind

The discovery document and jwks url can be setup one time using the storage account. This is the one I setup for our `azwi` CI:
- https://azwi.blob.core.windows.net/oidc-test/.well-known/openid-configuration
- https://azwi.blob.core.windows.net/oidc-test/openid/v1/jwks

The service account keys are in keyvault and retrieved when bootstrapping the kind cluster for the CI. xref: https://github.com/Azure/azure-workload-identity/blob/main/scripts/create-kind-cluster.sh

The federated identity credential can be setup single time if the namespace, service account name, namespace and OIDC issuer url is fixed.